src/modules/bulletproofs/inner_product_impl.h

```diff
@@ -0,0 +1,848 @@
+    }
+
+    return 1;
+}
+
+/* nb For security it is essential that `commit_inp` already commit to all data
+ *    needed to compute `P`. We do not hash it in during verification since `P`
+ *    may be specified indirectly as a bunch of scalar offsets.
+ */
+static int secp256k1_bulletproof_inner_product_verify_impl(const secp256k1_ecmult_context *ecmult_ctx, secp256k1_scratch *scratch, const secp256k1_bulletproof_generators *gens, size_t vec_len, const secp256k1_bulletproof_innerproduct_context *proof, size_t n_proofs, size_t plen, int shared_g) {
+    secp256k1_sha256 sha256;
+    secp256k1_bulletproof_innerproduct_vfy_ecmult_context ecmult_data;
+    unsigned char commit[32];
+    size_t total_n_points = 2 * vec_len + !!shared_g + 1; /* +1 for shared G (value_gen), +1 for H (blinding_gen) */
+    secp256k1_gej r;
+    secp256k1_scalar zero;
+    size_t i;
+
+    if (plen != secp256k1_bulletproof_innerproduct_proof_length(vec_len)) {
+        return 0;
+    }
+
+    if (n_proofs == 0) {
+        return 1;
+    }
+
+    if (!secp256k1_scratch_allocate_frame(scratch, n_proofs * (sizeof(*ecmult_data.randomizer) + sizeof(*ecmult_data.proof)), 2)) {
+        return 0;
+    }
+
+    secp256k1_scalar_clear(&zero);
+    ecmult_data.n_proofs = n_proofs;
+    ecmult_data.g = gens->blinding_gen;
+    ecmult_data.geng = gens->gens;
+    ecmult_data.genh = gens->gens + gens->n / 2;
+    ecmult_data.vec_len = vec_len;
+    ecmult_data.lg_vec_len = secp256k1_floor_lg(2 * vec_len / IP_AB_SCALARS);
+    ecmult_data.shared_g = shared_g;
+    ecmult_data.randomizer = (secp256k1_scalar *)secp256k1_scratch_alloc(scratch, n_proofs * sizeof(*ecmult_data.randomizer));
+    ecmult_data.proof = (secp256k1_bulletproof_innerproduct_vfy_data *)secp256k1_scratch_alloc(scratch, n_proofs * sizeof(*ecmult_data.proof));
+    /* Seed RNG for per-proof randomizers */
+    secp256k1_sha256_initialize(&sha256);
+    for (i = 0; i < n_proofs; i++) {
+        secp256k1_sha256_write(&sha256, proof[i].proof, plen);
+        secp256k1_sha256_write(&sha256, proof[i].commit, 32);
+        secp256k1_scalar_get_b32(commit, &proof[i].p_offs);
+        secp256k1_sha256_write(&sha256, commit, 32);
+    }
+    secp256k1_sha256_finalize(&sha256, commit);
+
+    secp256k1_scalar_clear(&ecmult_data.p_offs);
+    for (i = 0; i < n_proofs; i++) {
+        const unsigned char *serproof = proof[i].proof;
+        unsigned char proof_commit[32];
+        secp256k1_scalar dot;
+        secp256k1_scalar ab[IP_AB_SCALARS];
+        secp256k1_scalar negprod;
+        secp256k1_scalar x;
+        int overflow;
+        size_t j;
+        const size_t n_ab = 2 * vec_len < IP_AB_SCALARS ? 2 * vec_len : IP_AB_SCALARS;
+
+        total_n_points += 2 * ecmult_data.lg_vec_len + proof[i].n_extra_rangeproof_points - !!shared_g; /* -1 for shared G */
+
+        /* Extract dot product, will always be the first 32 bytes */
+        secp256k1_scalar_set_b32(&dot, serproof, &overflow);
+        if (overflow) {
```

Comment:
maybe `EXPECT` (not sure if it helps though). Same is true for all the other overflow checks.